neuromation/cli/object_storage.py

```diff
@@ -0,0 +1,367 @@
+            )
+        sources = *sources, destination
+        target_dir = parse_obj_or_file_resource(target_directory, root)
+        dst = None
+    else:
+        if destination is None:
+            raise click.MissingParameter(
+                param_type="argument", param_hint='"DESTINATION"'
+            )
+        if not sources:
+            raise click.MissingParameter(
+                param_type="argument", param_hint='"SOURCES..."'
+            )
+        dst = parse_obj_or_file_resource(destination, root)
+
+        # From gsutil:
+        #
+        # There's an additional wrinkle when working with subdirectories: the resulting
+        # names depend on whether the destination subdirectory exists. For example,
+        # if gs://my-bucket/subdir exists as a subdirectory, the command:
+
+        # gsutil cp -r dir1/dir2 gs://my-bucket/subdir
+
+        # will create the object gs://my-bucket/subdir/dir2/a/b/c. In contrast, if
+        # gs://my-bucket/subdir does not exist, this same gsutil cp command will create
+        # the object gs://my-bucket/subdir/a/b/c.
+        if no_target_directory or not await _is_dir(root, dst):
+            target_dir = None
+        else:
+            target_dir = dst
+            dst = None
+
+    filters = await calc_filters(root.client, filters)
+    srcs = await _expand(sources, root, glob, allow_file=True)
+    if no_target_directory and len(srcs) > 1:
+        raise click.UsageError(f"Extra operand after {str(srcs[1])!r}")
+
+    file_filter = FileFilter()
+    for exclude, pattern in filters:
+        file_filter.append(exclude, pattern)
+
+    show_progress = root.tty and progress
+
+    errors = False
+    for src in srcs:
+        if target_dir:
+            dst = target_dir / src.name
+        assert dst
+
+        progress_obj = create_storage_progress(root, show_progress)
+        progress_obj.begin(src, dst)
+
+        try:
+            if src.scheme == "file" and dst.scheme == "object":
+                if recursive and await _is_dir(root, src):
+                    await root.client.obj.upload_dir(
+                        src, dst, filter=file_filter.match, progress=progress_obj,
+                    )
+                else:
+                    await root.client.obj.upload_file(src, dst, progress=progress_obj)
+            elif src.scheme == "storage" and dst.scheme == "file":
+                raise NotImplementedError
+            else:
+                raise RuntimeError(
+                    f"Copy operation of the file with scheme '{src.scheme}'"
+                    f" to the file with scheme '{dst.scheme}'"
+                    f" is not supported"
+                )
+        except (OSError, ResourceNotFound, IllegalArgumentError) as error:
+            log.error(f"cannot copy {src} to {dst}: {error}")
+            errors = True
+
+        progress_obj.end()
+
+    if errors:
+        sys.exit(EX_OSFILE)
+
+
+async def _is_dir(root: Root, uri: URL) -> bool:
```

Comment:
Those 2 helper functions are behind a validation layer, could be refactored to be usable on both storages instead of a copy.